searx/network/client.py

```diff
@@ -155,56 +192,377 @@ def get_transport(verify, http2, local_address, proxy_url, limit, retries):
     )
 
 
-def new_client(
-    # pylint: disable=too-many-arguments
-    enable_http,
-    verify,
-    enable_http2,
-    max_connections,
-    max_keepalive_connections,
-    keepalive_expiry,
-    proxies,
-    local_address,
-    retries,
-    max_redirects,
-    hook_log_response,
-):
-    limit = httpx.Limits(
-        max_connections=max_connections,
-        max_keepalive_connections=max_keepalive_connections,
-        keepalive_expiry=keepalive_expiry,
-    )
-    # See https://www.python-httpx.org/advanced/#routing
-    mounts = {}
-    for pattern, proxy_url in proxies.items():
-        if not enable_http and pattern.startswith('http://'):
-            continue
-        if proxy_url.startswith('socks4://') or proxy_url.startswith('socks5://') or proxy_url.startswith('socks5h://'):
-            mounts[pattern] = get_transport_for_socks_proxy(
-                verify, enable_http2, local_address, proxy_url, limit, retries
+### Clients
+
+
+class ABCHTTPClient(ABC):
+    """Abstract HTTP client
+
+    Multiple implementation are defined bellow.
+    There are like an onion: each implementation relies on the previous one
+    and bring new feature.
+    """
+
+    @abstractmethod
+    def send(self, stream: bool, method: str, url: str, **kwargs) -> httpx.Response:
+        pass
+
+    @abstractmethod
+    def close(self):
+        pass
+
+    @property
+    @abstractmethod
+    def is_closed(self) -> bool:
+        pass
+
+    def request(self, method, url, **kwargs) -> httpx.Response:
+        return self.send(False, method, url, **kwargs)
+
+    def stream(self, method, url, **kwargs) -> httpx.Response:
+        return self.send(True, method, url, **kwargs)
+
+
+class OneHTTPClient(ABCHTTPClient):
+    """Wrap a httpx.Client
+
+    Use httpx_socks for socks proxies.
+
+    Deal with httpx.RemoteProtocolError exception: httpx raises this exception when the
+    HTTP/2 server disconnect. It is excepted to reconnect.
+    Related to https://github.com/encode/httpx/issues/1478
+    Perhaps it can be removed now : TODO check in production.
+
+    To be backward compatible with Request:
+
+        * In Response, "ok" is set to "not response.is_error()"
+          See https://www.python-httpx.org/compatibility/#checking-for-success-and-failure-responses
+
+        * allow_redirects is accepted
+          See https://www.python-httpx.org/compatibility/#redirects
+    """
+
+    def __init__(
+        # pylint: disable=too-many-arguments
+        self,
+        verify=True,
+        enable_http=True,
+        enable_http2=False,
+        max_connections=None,
+        max_keepalive_connections=None,
+        keepalive_expiry=None,
+        proxies=None,
+        local_addresses=None,
```

Comment:
@dalf : Its not entirely clear to me, does a `HTTPClient` really have more than on `local_addresses`? .. I know the `Network` can have more than on local address .. but the `HTTPCLient`?

Reply:
Ah good point the parameter name should be `local_address` not `local_addresses` :
* `HTTPClient` have only one `local_addresses`
* `Network` have a mapping between [`(local_address, proxies)`](https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/network.py#L139-L140) and an `HTTPClient`.

Reply:
> * `Network` have a mapping between [`(local_address, proxies)`](https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/network.py#L139-L140) and an `HTTPClient`.

yes, this was the point where I struggled :) .. I will fix the naming in my reviewhttps://github.com/return42/searxng/commits/searx_network_rework